Allow enforcing HTTPS for custom domains

davidfischer opened the pull request:
This allows users to mark a domain as HTTPS and then all links to its documentation will be HTTPS links instead of HTTP links.

Personally I don't love these changes to the resolver but I believe the resolver needs a larger refactor for performance reasons anyway (https://github.com/rtfd/readthedocs.org/issues/3712). If there's a better way to add this functionality without adding more SQL queries and degrading performance of a very performance sensitive bit of code, please let me know or feel free to add to this.

Fixes https://github.com/rtfd/readthedocs.org/issues/2236

## Screenshot

<img width="882" alt="screen shot 2018-08-06 at 3 17 06 pm" src="https://user-images.githubusercontent.com/185043/43743726-dae0c240-998b-11e8-990a-abf5c25f4e6a.png">

davidfischer commented on readthedocs/projects/models.py:
```diff
@@ -985,7 +985,7 @@ class Domain(models.Model):
         _('Use HTTPS'),
         default=False,
-        help_text=_('SSL is enabled for this domain')
+        help_text=_('Always use HTTPS for this domain')
```
Technically this will result in a migration. However, there's a bunch of other minor pending migrations on `Project`.

humitos commented on readthedocs/core/resolver.py:
```diff
@@ -144,13 +144,24 @@ def resolve(self, project, protocol='http', filename='', private=None,
+        elif self._use_subdomain():
+            domain = self._get_project_subdomain(canonical_project)
+        else:
+            domain = getattr(settings, 'PRODUCTION_DOMAIN')
```
Isn't this `if/elif/else` the same code of what `self.resolve_domain` does?

davidfischer replied:
It is. This function needs access to the `Domain` object in order to read the `.https` property. That object is not exposed from the `resolve_domain` method.

davidfischer replied:
Loading that object again based on the hostname alone will require another database query and the resolver is already the largest single source of database load in Read the Docs.

humitos replied:
I understand and I saw that you mentioned the issue to refactor this. So, I'm fine with "duplication" here.

Although, I think that adding a similar explanation written in these comments as a comment in the code would make it easier to refactor this in the future.

davidfischer commented:
> Besides the tests you changed, shouldn't we have a specific test that uses a Domain(https=False) and check that it resolves to http?

Probably yes.

davidfischer commented:
I updated based on the feedback to add another resolver test and to comment on the duplication.

I also added a UI element for showing the status of issuing a certificate.
